Copy-edit (#88)
ember-learn/super-rentals-tutorial: src/chapters/14-provider-components.md

In order to make our search box actually work, we are going to need to retain and store the text that the user types in when they use the search box. This text is the search query, and it is a piece of *[state][TODO: link to state]* that is going to change whenever the user types something into the search box.

But where are we going to put this newly-introduced piece of state? In order to wire up the search box, we need a place to store the search query. At the moment, our search box lives on the `index.hbs` route template, which doesn't have a good place to store this search query state. Darn, this would be so much easier to do if we had a component, because we could just store the state directly on the component!

> Wait...why don't we just refactor the search box into a component? Once we do that, this will all be a bit easier&mdash;hooray!

Reviewer: Something something spaces around emdash (I think we settled on spaces around it)

Reviewer 2: I ended up [changing it](https://github.com/ember-learn/super-rentals-tutorial/pull/75/commits/d6fec45a20c626f347a2d880f8cb87ced7689a24) to have no spaces because it's actually...grammatically incorrect to do that. I kept spaces because it was easier to read while we were writing the prose...but now I changed it 😬

</div>
```

> There is one minor change to note here: while extracting our markup into a component, we also renamed the `@model` argument to be `@rentals` instead, just in order to be a little more specific about what we're iterating over in our `#each` loop. Otherwise, all we're doing here is copy-pasting what was on our `index.hbs` page into our new component template. Now we just need to actually use our new component in the index template where we started this whole refactor! Let's render our `<Rentals>` component in our `index.hbs` template.

Reviewer: ```suggestion
There is one minor change to note here: while extracting our markup into a component, we also renamed the `@model` argument to be `@rentals` instead, just in order to be a little more specific about what we're iterating over in our `{{#each}}` loop. Otherwise, all we're doing here is copy-pasting what was on our `index.hbs` page into our new component template. Now we just need to actually use our new component in the index template where we started this whole refactor! Let's render our `<Rentals>` component in our `index.hbs` template.
```

I think that's how are referring to it in the other chapters.

Reviewer 2: I don't get this suggestion...looks like nothing changed?

Reviewer 2: oh i see it now. these suggestion things make it so hard to difficult to parse small diffs 😢

```run:file:patch lang=handlebars cwd=super-rentals filename=app/templates/index.hbs
@@ -6,13 +6,2 @@

-<div class="rentals">
-  <label>
-    <span>Where would you like to stay?</span>
-    <input class="light">
-  </label>
-
-  <ul class="results">
-    {{#each @model as |rental|}}
-      <li><Rental @rental={{rental}} /></li>
-    {{/each}}
-  </ul>
-</div>
+<Rentals @rentals={{@model}} />
```

Remember the small change we made in the markup when we extracted our `<Rentals>` component? We renamed the `@model` argument to be `@rentals`. Because we made that change in our component, we now need to pass the `@model` argument into the `<Rentals>` component as `@rentals`. Once we do this, everything should be wired up properly so that the `@model` is passed into `<Rentals>` as `@rentals`, just as we expect.

Let's check our UI as well to make sure that we didn't break anything during this refactor...

> ```run:screenshot width=1024 retina=true filename=homepage-with-rentals-component.png alt="The homepage looks exactly the same as before!"

Reviewer 2: In the current version of this chapter, we have a `yarn test` and `git add`, but the copy edits in this PR move things around a bit. I'll make sure to check that those two commands exist before the screenshot and add all your suggestions into a single commit after this PR is merged 👍

wait  #qunit-banner.qunit-pass
```

> ## Using Ember's `<Input>`

Reviewer: ```suggestion
## Using Ember's `<Input>` Component
```

git add app/components/rentals.js
```

> ## Adding the <Rentals::Filter> Provider Component

Reviewer: ```suggestion
## Adding the `<Rentals::Filter>` Provider Component
```

(Either way seems fine to me, but seems like this is what we have been doing)

Now that our search query is wired up to our `<Rentals>` component, we can get into the really fun stuff! Namely, we can make our component *filter* results based on our search query. In order to encapsulate this functionality, we'll create another component called `<Rentals::Filter>`.

> 

Reviewer: Can't comment on the folded part below, but:

> Instead, we're `yield`ing to something, using the `{{yield}}` keyword...

👇 

> Instead, we're yielding to something, using the `{{yield}}` keyword

Reviewer: Also:

> ...let's take a look at how we use the `{{yield}}` keyword in the `<Rentals>` component.

This sentence is a bit confusing, in that, the code snippet immediately below it doesn't actually contains a `{{yield}}`.

However, the main difference here is the use of `as |results|` when we are invoking our `<Rentals::Filter>` component. Incidentally, this new syntax goes hand-in-hand with the `{{yield this.results}}` syntax we were introduced to in the component template.

The `as |results|` syntax might look a little new to us, but it isn't the first time that we've seen this feature in action. Back when we first learned about the `{{#each}}` syntax, which we use to loop over a collection, we wrote something like this: `{{#each @items as |item|}}...some content here...{{/each}}`.

When we use this syntax, we are passing a block&mdash;the `...some content here...` in our example&mdash;to `{{#each}}`. Ember will iterate through the array we provided (`@items`) and render our block _once per item_ in the array.

> Inside of our block, we need to be able to access the current item _somehow_. The `{{#each}}` helper provides the item to our block via the `as |item|` declaration, which creates a local variable `item`, also known as a *[block parameter][TODO: link to block parameter]*. In other words, as we iterate through `@items`, we will have access to the current item that we're looping over through the block parameter (`item`). The block parameter is only accessible from inside the block. Ember will fill in the block parameter with the current item of the iteration, and it will do this each time that it calls our block.

Reviewer: ```suggestion
Inside of our block, we need to be able to access the current item _somehow_. The `{{#each}}` syntax provides the item to our block via the `as |item|` declaration, which creates a local variable `item`, also known as a *[block parameter][TODO: link to block parameter]*. In other words, as we iterate through `@items`, we will have access to the current item that we're looping over through the block parameter (`item`). The block parameter is only accessible from inside the block. Ember will fill in the block parameter with the current item of the iteration, and it will do this each time that it renders our block.
```

The need to provide some data to a block is not unique to the `{{#each}}` syntax. In this case, our `<Rentals::Filter>` component wants to take the unfiltered list of rental properties and match them against the user's query. Once the component has matched the rentals against the query, it will need to provide a filtered list of rental properties to its caller (the `<Rentals>` component).

> 

Reviewer 2: seems good

> Zoey says...
>

> > The local variable name `results` is arbitrary, and isn't special in any way! You could name it anything: `as |data|`, `as |filtered|`, or even `as |banana|`! Really, the important thing here is that however you name the block param is how you will have access to the yielded data from inside the block.

Reviewer 2: good idea, i like drawing that comparison!

Interestingly, if we take a look at our `<Rentals::Filter>` component template, we see that we don't actually render any content. Instead, this component's only responsibility is to set up some piece of state (`this.results`, the list of filtered rental properties), and then yield that state back up to its caller (`<Rentals>`) in the form of a block parameter (`as |results|`).

> This is called the (provider component pattern)[TODO: link to provider component pattern], which we see in action with one component providing data up to its caller.

Reviewer: ```suggestion
This is called the *(provider component pattern)[TODO: link to provider component pattern]*, which we see in action with one component providing data up to its caller.
```

Something about the "..., which we see in action..." part reads a bit confusing to me. 🤔 I think it may be fine to just drop that second half and merge the sentence into the previous paragraph?

 });
```

Great! In the process of adding this test, we'll notice that we also extracted our setup (`setProperties`) into the before hooks. We also used the `fillIn` test helper in our newly-added test.

> Zoey says...
>

> > This search functionality is not perfect. Ideally, it would also be case-insensitive and allow you to search by city, category, type, or description. If you're looking for a challenge, see if you can improve on our search!

Reviewer: ```suggestion
> This search functionality is not perfect. Ideally, it would also be case-insensitive, and also allow you to search by city, category, type, or description. If you're looking for a challenge, see if you can improve on our search!
```
